src/Sarif.FunctionalTests/ConverterTestData/StaticDriverVerifier/checkadddevice_bug1.tt.sarif

```diff
@@ -32,154 +32,105 @@
             {
               "locations": [
                 {
+                  "fullyQualifiedLogicalName": "OS",
                   "kind": "call",
+                  "callee": "main",
                   "importance": "unimportant",
                   "properties": {
-                    "State": "^",
-                    "Caller": "OS",
-                    "Callee": "main"
+                    "currentDataValues": "^"
                   }
                 },
                 {
                   "step": 1,
-                  "message": "====Auto=====",
                   "kind": "continuation",
-                  "importance": "unimportant",
-                  "properties": {
-                    "State": "^====Auto====="
-                  }
+                  "importance": "unimportant"
                 },
                 {
                   "step": 2,
-                  "message": "====Auto=====,====Pre=====,____true",
+                  "fullyQualifiedLogicalName": "main",
                   "kind": "call",
+                  "callee": "_sdv_init4",
                   "importance": "unimportant",
                   "properties": {
-                    "State": "^====Auto=====^====Pre=====^____true",
-                    "Caller": "main",
-                    "Callee": "_sdv_init4"
+                    "currentDataValues": "^====Pre=====^____true"
                   }
                 },
                 {
                   "step": 3,
-                  "message": "====Auto=====",
                   "kind": "continuation",
-                  "importance": "unimportant",
-                  "properties": {
-                    "State": "^====Auto====="
-                  }
+                  "importance": "unimportant"
                 },
                 {
                   "step": 4,
-                  "message": "====Auto=====",
                   "kind": "assignment",
-                  "importance": "unimportant",
-                  "properties": {
-                    "State": "^====Auto====="
-                  }
+                  "importance": "unimportant"
                 },
                 {
                   "step": 5,
-                  "message": "====Auto=====",
                   "kind": "assignment",
-                  "importance": "unimportant",
-                  "properties": {
-                    "State": "^====Auto====="
-                  }
+                  "importance": "unimportant"
                 },
                 {
                   "step": 6,
-                  "message": "====Auto=====",
+                  "fullyQualifiedLogicalName": "main",
```

Comment:
Re: FQLN for `callReturn`: OK, I agree with your argument that since we also have `functionEnter` and `functionExit`, then `functionExit` should refer to the actual `return` statement in the callee. That in turn implies that `callReturn` refers to a location in the caller.

So there are two questions:

1. In a .tt file, does a "Return" line (for example, `269 "?" 0 true ^====Auto=====^s=1 Return`) really mean a `return` statement in the code (in which case, SARIF should represent it with `kind = functionExit`), or does it really mean the next location in the caller (in which case SARIF should represent it with `kind = callReturn`)? I will ask Jakob.

2. When the viewer builds the tree, it _indents_ the `callReturn` aCLs under the caller. By your argument, the `callReturn` should pop back to the same indentation level as the caller, right?

Re: Call/callReturn pairs: the spec does not require them to occur in matched pairs. Since a codeFlow is a possible path through the code (generally used to specify a code path leading to the discovery of a problem), it's perfectly fine for the code path to be:

- a calls b
- b calls c
- c calls d
- BOOM